### PR TITLE
Biconnectivity module using visitor functions

### PR DESCRIPTION
**Pull request by abhinavmehndiratta**
**Benchmarks** 

There is a slowdown of about 20%, but I'm trying to fix that.

New -
```julia
julia> g = loadsnap(:facebook_combined)
{4039, 88234} undirected simple Int64 graph

julia> @btime LightGraphs.Biconnectivity.articulation(g)
  1.441 ms (28 allocations: 192.73 KiB)

julia> @btime LightGraphs.Biconnectivity.bridges(g)
  1.463 ms (29 allocations: 195.95 KiB)

julia> @btime LightGraphs.Biconnectivity.biconnected_components(g)
  4.205 ms (248 allocations: 4.59 MiB)
```
Old -
```julia

julia> g1 = loadsnap(:facebook_combined)
{4039, 88234} undirected simple Int64 graph

julia> @btime articulation(g)
  1.135 ms (22 allocations: 160.92 KiB)

julia> @btime bridges(g)
  1.170 ms (22 allocations: 163.98 KiB)

julia> @btime biconnected_components(g)
  3.782 ms (231 allocations: 4.46 MiB)
```

### CODE REVIEW

**Comment by sbromberger** [comment]
As discussed - we should make sure the edges are ordered.

**Comment by sbromberger** [comment]
Also - are those benchmarks you posted using the incorrect old version of biconnected_components, or the fixed version?

**Comment by abhinavmehndiratta** [comment]
> Also - are those benchmarks you posted using the incorrect old version of biconnected_components, or the fixed version?

The fixed version. Also, after merging the new visitor return codes, this PR will need some changes though (changing `return true` to `return VSUCCESS`).

**Comment by abhinavmehndiratta** [comment]
> As discussed - we should make sure the edges are ordered.

Okay, I'll do that. But still need to figure out how to make this fast.

**Comment by sbromberger** [comment]
Yeah. I'm concerned that this is slower than the fixed version, which was itself slower than the incorrect version.

File: src/Biconnectivity/articulation.jl [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+3-element Array{Int64,1}:
+ 2
+ 3
+ 4
+```
+"""
+function articulation end
+@traitfn function articulation(g::AG::(!IsDirected)) where {T, AG<:AbstractGraph{T}}
+    state = ArticulationState(T(nv(g)))
+    for u in vertices(g)
```
Comment (sbromberger) [inline src/Biconnectivity/articulation.jl:74]: `@inbounds` here might help.

File: src/Biconnectivity/articulation.jl [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            state.nchildren = 0
+            state.prnt[u] = 0
+            traverse_graph!(g, u, DepthFirst(), state)
+            if state.nchildren > 1
+                state.is_ap[u] = true
+            end
+        end
+    end
+    articulation_points = Vector{T}()
+    for u in vertices(g)
```
Comment (sbromberger) [inline src/Biconnectivity/articulation.jl:87]: `@inbounds`

File: src/Biconnectivity/biconnect.jl [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+julia> biconnected_components(cycle_graph(5))
+1-element Array{Array{LightGraphs.SimpleGraphs.SimpleEdge,1},1}:
+ [Edge 1 => 5, Edge 4 => 5, Edge 3 => 4, Edge 2 => 3, Edge 1 => 2]
+```
+"""
+function biconnected_components end
+@traitfn function biconnected_components(g::AG::(!IsDirected)) where {T, AG<:AbstractGraph{T}}
+    state = BiconnectState(T(nv(g)))
+    for u in vertices(g)
```
Comment (sbromberger) [inline src/Biconnectivity/biconnect.jl:95]: `@inbounds`?

File: src/Biconnectivity/bridge.jl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ Edge 4 => 5
+ Edge 3 => 4
+ Edge 2 => 3
+ Edge 1 => 2
+```
+"""
+function bridges end
+@traitfn function bridges(g::AG::(!IsDirected)) where {T, AG<:AbstractGraph{T}}
+    state = BridgeState(T(nv(g)))
+    for u in vertices(g)
```
Comment (sbromberger) [inline src/Biconnectivity/bridge.jl:74]: `@inbounds`?

**Comment by abhinavmehndiratta** [comment]
Closing this, refer #1446

**Pull request closed without merging**